{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Haplotyper Overview\n",
    "\n",
    "### Input\n",
    "  - gene definition file\n",
    "  - per-barcode variants file \n",
    "  \n",
    "  \n",
    "### For each allele\n",
    "  - make a set containing the allelic variants\n",
    "  - classify each variant as either known or novel\n",
    "  - compare the set of allele variants with the variants for each defined haplotype\n",
    "    - set of variants shared between allele and haplotype\n",
    "    - fraction of haplotype variants observed in allele\n",
    "    - Jaccard coefficient of the allele variants and haplotype variants\n",
    "    - Number of 'significant' variants for the haplotype present in the allele sequence\n",
    "\n",
    "### Filter and sort the results\n",
    "  - ignore any haplotypes which do not have snps\n",
    "  - sort by jaccard, fraction and number of 'significant' variants\n",
    "    - best 'scoring' haplotypes will be at the head of the list\n",
    "\n",
    "### Output\n",
    "  - all matches and summary information in json format for further processing\n",
    "  - a tabular summary for humans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json, yaml\n",
    "\n",
    "# load reference data\n",
    "with open(\"../../../CYP2D6_Pipeline/test_data/gene/CYP2D6.yaml\", \"r\") as infile:\n",
    "    gene = yaml.safe_load(infile)\n",
    "    \n",
    "# make a set containing all reference variants for the gene\n",
    "reference_variants = {gene[\"snps\"][snp][\"g_notation\"] for snp in gene[\"snps\"]}\n",
    "       \n",
    "def characterise_allele(found_variants, gene_definition):\n",
    "    \"\"\"\n",
    "    Compare the variants for an allele with all the known haplotypes.\n",
    "    Characterise the commonality between the allele and the reference haplotypes.\n",
    "    \n",
    "    :param found_variants: A set of variants observed in the allele sequence\n",
    "    :type found_variants: {str, str, ...} where str is a string containing an hgvs variant description\n",
    "    \n",
    "    :param gene_definition: The parsed gene file containing snp and haplotype definitions\n",
    "    :type gene_definition: dict\n",
    "    \n",
    "    :return: a list of dicts where each dict characterises the commonality between the allelic\n",
    "             variants and the variants for a single haplotype\n",
    "    \"\"\"\n",
    "    # list to hold the comparisons \n",
    "    summary = []\n",
    "    \n",
    "    # compare the allele variants with those for each defined haplotype\n",
    "    for haplotype in (h for h in gene_definition[\"haplotypes\"] if len(h[\"snps\"]) > 0):\n",
    "        # the variants for this haplotype \n",
    "        haplotype_variants = {gene_definition[\"snps\"][s][\"g_notation\"] for s in haplotype[\"snps\"]}\n",
    "        \n",
    "        # the variants shared between the allele and the haplotype\n",
    "        shared_variants = found_variants & haplotype_variants\n",
    "        \n",
    "        # summarize\n",
    "        try:\n",
    "            summary.append(\n",
    "                {\n",
    "                    # the current haplotype\n",
    "                    \"haplotype\": haplotype[\"type\"], \n",
    "                    \n",
    "                    # the variants for this haplotype\n",
    "                    \"haplotype_variants\": sorted(list(haplotype_variants)), \n",
    "                    \n",
    "                    # the variants shared between allele and haplotype\n",
    "                    \"shared_variants\": sorted(list(shared_variants)), \n",
    "                    \n",
    "                    # the fraction of haplotype variants observed\n",
    "                    \"fraction\": len(shared_variants)/ len(haplotype_variants),\n",
    "                    \n",
    "                    # jaccard coefficient of haplotype and allele variants\n",
    "                    \"jaccard\": len(shared_variants) / (len(haplotype_variants) + len(found_variants) - len(shared_variants)),\n",
    "                   \n",
    "                    # shared 'significant' variants\n",
    "                    \"significant\": [\n",
    "                        gene_definition[\"snps\"][s][\"g_notation\"] for s in haplotype[\"snps\"] if gene_definition[\"snps\"][s][\"tags\"] is not None and \\\n",
    "                        \"significant\" in gene_definition[\"snps\"][s][\"tags\"] and \\\n",
    "                        gene_definition[\"snps\"][s][\"g_notation\"] in shared_variants\n",
    "                    ]\n",
    "                }\n",
    "            )\n",
    "        except ZeroDivisionError:\n",
    "            pass\n",
    "    \n",
    "    return summary\n",
    "\n",
    "\n",
    "def match_allele(variants, gene, trim_boundary=False):\n",
    "    \"\"\"\n",
    "    Compare the allele variants against the haplotype definitions\n",
    "    Return a dictionary summarizing the comparison\n",
    "    \"\"\"\n",
    "    # sort the variants\n",
    "    variants = sorted(variants)\n",
    "    \n",
    "    # trim the first and last variants if requested\n",
    "    if trim_boundary:\n",
    "        if len(variants) > 0 and \"ins\" in variants[0]:\n",
    "            variants = variants[1:]\n",
    "    \n",
    "        if len(variants) > 0  and \"ins\" in variants[-1]:\n",
    "            variants = variants[:-1]\n",
    "    \n",
    "    # set of variants found in the allele\n",
    "    found_variants = set(variants)\n",
    "    \n",
    "    # known variants - intersection of found variants with the set of all reference variants\n",
    "    known_variants = found_variants & reference_variants\n",
    "    \n",
    "    # novel variants found in allele but not found in the reference set\n",
    "    novel_variants = found_variants - reference_variants\n",
    "    \n",
    "    # significant variants found in allele\n",
    "    significant_variants = {\n",
    "        gene[\"snps\"][s][\"g_notation\"] for s in gene[\"snps\"] if gene[\"snps\"][s][\"tags\"] is not None and \\\n",
    "        \"significant\" in gene[\"snps\"][s][\"tags\"] and \\\n",
    "        gene[\"snps\"][s][\"g_notation\"] in found_variants\n",
    "    }\n",
    "\n",
    "    # get all haplotype matches for this allele\n",
    "    characterised_allele = characterise_allele(found_variants, gene)\n",
    "    \n",
    "    # filter to remove matches with zero overlap\n",
    "    filtered_allele = [x for x in characterised_allele if x[\"fraction\"] > 0]\n",
    "    \n",
    "    # sort by number of significant variants, fraction and jaccard\n",
    "    # this gives a list of haplotypes ordered by decreasing 'score'\n",
    "    sorted_allele = sorted(filtered_allele, key=lambda x: (len(x[\"significant\"]), x[\"fraction\"], x[\"jaccard\"]), reverse=True)\n",
    "    \n",
    "    return {\n",
    "        \"known_variants\": sorted(list(known_variants)),\n",
    "        \"novel_variants\": sorted(list(novel_variants)),\n",
    "        \"significant_variants\": sorted(list(significant_variants)),\n",
    "        \"haplotypes\":     sorted_allele\n",
    "    }\n",
    "\n",
    "\n",
    "def tabulate_allele_matches(allele, match, num_hits=None):\n",
    "    \"\"\"\n",
    "    Convert allele match information into a tabular representation for printing\n",
    "    \"\"\"\n",
    "    overview = \"\\n\".join(\n",
    "        [\n",
    "            \"- {}    %id: {}\".format(allele[\"sequence_id\"], allele[\"identity\"]),\n",
    "            \"- Variants: {}    Known: {}    Novel: {}    Significant: {}\".format(\n",
    "                len(match[\"known_variants\"]) + len(match[\"novel_variants\"]),\n",
    "                len(match[\"known_variants\"]),\n",
    "                len(match[\"novel_variants\"]),\n",
    "                len(match[\"significant_variants\"])\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "    \n",
    "    report_hits = match[\"haplotypes\"][:num_hits]\n",
    "    columns = []\n",
    "    columns.append([\"{:25}\".format(\"Variant|Haplotype\")] + \\\n",
    "                   [\"\\033[1m{:25}\\033[0m\".format(v) if v in match[\"significant_variants\"] else \"{:25}\".format(v) for v in match[\"known_variants\"]] + \\\n",
    "                   [\"{:25}\".format(label) for label in [\"Fraction\", \"Proportion\", \"Jaccard\", \"Significant\"]])\n",
    "    \n",
    "    for hit in report_hits:\n",
    "        column = [\"{:6}\".format(hit[\"haplotype\"])]\n",
    "        column += [\"{:6}\".format(\"*\" if v in hit[\"haplotype_variants\"] else \" \") for v in match[\"known_variants\"]]\n",
    "        column.append(\"{:6}\".format(\"{}/{}\".format(len(hit[\"shared_variants\"]), len(hit[\"haplotype_variants\"]))))\n",
    "        column.append(\"{:6}\".format(str(hit[\"fraction\"])[:4]))\n",
    "        column.append(\"{:6}\".format(str(hit[\"jaccard\"])[:4]))\n",
    "        column.append(\"{:<6}\".format(len(hit[\"significant\"])))\n",
    "        columns.append(column)\n",
    "        \n",
    "    results = []\n",
    "    for row in range(len(columns[0])):\n",
    "        results.append(\"\".join([c[row] for c in columns]))\n",
    "    \n",
    "    table = \"\\n\".join(results)\n",
    "    \n",
    "    return (overview, table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- Barcodelbc66--lbc66_Cluster0_Phase1_NumReads245    %id: 0.9843298341700898\n",
      "- Variants: 29    Known: 23    Novel: 6    Significant: 1\n",
      "Variant|Haplotype        *35A  *35   *35XN *35B  *2A   *2    *2XN  *2D   *39   *34   *31   *56A  *105  *98   *104  *21B  *51   *88   *102  *41   *41X2 *2M   *14B  *21   *65   \n",
      "42126310C>T                                *                                         *           *           *                       *                 *                 *     \n",
      "42126611C>G              *     *     *     *     *     *     *     *     *           *     *     *     *     *     *     *     *     *     *     *     *     *     *     *     \n",
      "42127001G>A                                *                                         *     *     *     *     *                 *     *                 *                 *     \n",
      "42127207C>T                                *                                         *     *           *                       *                       *                 *     \n",
      "42127407T>G                                *                                         *     *     *     *     *                 *     *                 *                 *     \n",
      "42127941G>A              *     *     *     *     *     *     *     *           *     *     *     *     *     *     *     *           *     *     *     *     *     *     *     \n",
      "42129130C>G              *     *     *     *     *     *     *           *           *     *     *     *     *     *     *     *     *     *     *     *     *     *     *     \n",
      "42129950A>C                                *                                         *           *     *     *                 *                       *                 *     \n",
      "42130047G>C                                *                                         *           *     *                       *                       *                       \n",
      "42130482C>A                                *                                         *           *           *                       *                 *                 *     \n",
      "42130547T>C                                      *                                   *     *     *     *     *     *     *     *     *     *     *     *     *     *           \n",
      "42130559T>G                                      *                                   *     *     *     *     *     *     *     *     *     *     *     *     *     *           \n",
      "42130560C>G                                      *                                   *     *     *     *     *     *     *     *     *     *     *     *     *     *           \n",
      "42130565A>G                                      *                                   *     *     *     *     *     *     *     *     *     *     *     *     *     *           \n",
      "42130569G>C                                      *                                   *     *     *     *     *     *     *     *     *     *     *     *     *     *           \n",
      "42130571G>T                                      *                                   *     *     *     *     *     *     *     *     *     *     *     *     *     *           \n",
      "42130578C>G                                      *                                   *     *     *     *     *     *     *     *     *     *     *     *     *     *           \n",
      "\u001b[1m42130761C>T              \u001b[0m*     *     *     *                                                                                                                                   \n",
      "42131469C>T                                *     *                                   *     *                       *     *                 *     *     *                       \n",
      "42131531G>A                                *     *                                   *     *                       *     *                 *     *     *                       \n",
      "42132026T>C                                *     *                                   *     *                       *     *                 *     *     *                       \n",
      "42132375G>C              *                 *     *                                   *     *                       *     *                                                     \n",
      "42132561C>T                                *                                         *                                                                                         \n",
      "Fraction                 5/5   4/4   4/4   16/32 14/14 3/3   3/3   2/2   2/2   1/1   22/23 17/18 16/17 15/16 15/16 14/15 14/15 14/15 14/15 13/14 13/14 20/22 10/11 10/11 9/10  \n",
      "Proportion               1.0   1.0   1.0   0.5   1.0   1.0   1.0   1.0   1.0   1.0   0.95  0.94  0.94  0.93  0.93  0.93  0.93  0.93  0.93  0.92  0.92  0.90  0.90  0.90  0.9   \n",
      "Jaccard                  0.17  0.13  0.13  0.35  0.48  0.10  0.10  0.06  0.06  0.03  0.73  0.56  0.53  0.5   0.5   0.46  0.46  0.46  0.46  0.43  0.43  0.64  0.33  0.33  0.3   \n",
      "Significant              1     1     1     1     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     \n",
      "\n",
      "- Barcodelbc66--lbc66_Cluster0_Phase2_NumReads217    %id: 0.9846329402814759\n",
      "- Variants: 28    Known: 23    Novel: 5    Significant: 1\n",
      "Variant|Haplotype        *41   *41X2 *69   *91   *2    *2XN  *2D   *39   *34   *2M   *105  *98   *104  *88   *102  *2A   *31   *14B  *21   *65   *73   *56A  *103  *21B  *51   \n",
      "42126310C>T                          *                                         *     *           *           *           *                 *                 *                 \n",
      "42126611C>G              *     *     *           *     *     *     *           *     *     *     *     *     *     *     *     *     *     *     *     *     *     *     *     \n",
      "42127001G>A                          *                                         *     *     *     *     *     *           *                 *     *     *     *                 \n",
      "42127207C>T                          *                                         *           *           *                 *                 *     *     *                       \n",
      "42127407T>G                          *                                         *     *     *     *     *     *           *                 *     *     *     *                 \n",
      "\u001b[1m42127803C>T              \u001b[0m*     *     *     *                                                                                                                                   \n",
      "42127941G>A              *     *     *     *     *     *     *           *     *     *     *     *           *     *     *     *     *     *     *     *     *     *     *     \n",
      "42129130C>G              *     *     *           *     *           *           *     *     *     *     *     *     *     *     *     *     *     *     *     *     *     *     \n",
      "42129950A>C                          *                                         *     *     *     *     *                 *                 *     *                             \n",
      "42130047G>C                          *                                         *     *     *           *                 *                       *                             \n",
      "42130482C>A                          *                                         *     *           *           *           *                 *     *           *                 \n",
      "42130547T>C              *     *                                               *     *     *     *     *     *     *     *     *     *           *     *     *     *     *     \n",
      "42130559T>G              *     *                                               *     *     *     *     *     *     *     *     *     *           *     *     *     *     *     \n",
      "42130560C>G              *     *                                               *     *     *     *     *     *     *     *     *     *           *     *     *     *     *     \n",
      "42130565A>G              *     *                                               *     *     *     *     *     *     *     *     *     *           *     *     *     *     *     \n",
      "42130569G>C              *     *                                               *     *     *     *     *     *     *     *     *     *           *     *     *     *     *     \n",
      "42130571G>T              *     *                                               *     *     *     *     *     *     *     *     *     *           *     *     *     *     *     \n",
      "42130578C>G              *     *                                               *     *     *     *     *     *     *     *     *     *           *     *     *     *     *     \n",
      "42131469C>T              *     *                                               *                                   *     *                             *           *     *     \n",
      "42131531G>A              *     *                                               *                                   *     *                       *     *           *     *     \n",
      "42132026T>C              *     *     *                                         *                                   *     *                             *           *     *     \n",
      "42132049_42132050insTT                                                         *                                                                                               \n",
      "42132561C>T                                                                                                              *                                                     \n",
      "Fraction                 14/14 14/14 12/17 2/3   3/3   3/3   2/2   2/2   1/1   21/22 16/17 15/16 15/16 14/15 14/15 13/14 21/23 10/11 10/11 9/10  17/19 16/18 14/16 13/15 13/15 \n",
      "Proportion               1.0   1.0   0.70  0.66  1.0   1.0   1.0   1.0   1.0   0.95  0.94  0.93  0.93  0.93  0.93  0.92  0.91  0.90  0.90  0.9   0.89  0.88  0.87  0.86  0.86  \n",
      "Jaccard                  0.5   0.5   0.36  0.06  0.10  0.10  0.07  0.07  0.03  0.72  0.55  0.51  0.51  0.48  0.48  0.44  0.7   0.34  0.34  0.31  0.56  0.53  0.46  0.43  0.43  \n",
      "Significant              1     1     1     1     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     0     \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# load barcode variants and display\n",
    "with open(\"../../variant_calling/96_well_output/variants/lbc66.json\", \"r\") as infile:\n",
    "    alleles = json.load(infile)\n",
    "\n",
    "results = []\n",
    "for allele in alleles:\n",
    "    match = match_allele(allele[\"variants\"], gene)\n",
    "    match[\"sequence_id\"] = allele[\"sequence_id\"]\n",
    "    \n",
    "    table = tabulate_allele_matches(allele, match, 25)\n",
    "    \n",
    "    results.append(\n",
    "        (\n",
    "            json.dumps(match, indent=4, separators=(',', ': ')),\n",
    "            \"\\n\".join(table)\n",
    "        )\n",
    "    )\n",
    "\n",
    "\n",
    "for result in results:\n",
    "    print(result[1])\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:95% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%html\n",
    "<style>.container { width:95% !important; }</style>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
